File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer tokens;
	
	static int N; // N*N Map 크기
	static int M; // M개 빼곤 치킨집 다 폐업
	static List<int[]> Home = new ArrayList<>(); // 집 좌표
	static List<int[]> Chicken = new ArrayList<>(); // 치킨집 좌표
	static int[][] dis; // 치킨집-집 간의 거리
	static int Min = Integer.MAX_VALUE;
	
	public static void combi(int nth, int[] choosed, int start) {
		if(nth == M) {
			int sum = 0;
			for(int i=0; i<Home.size(); i++) {
				int min = Integer.MAX_VALUE;
				for(int j=0; j<choosed.length; j++) {
					min = Math.min(min, dis[choosed[j]][i]);
				}
				sum += min;
			}
			Min = Math.min(Min, sum);
			return;
		}
		for(int i=start; i<Chicken.size(); i++) {
			choosed[nth] = i;
			combi(nth + 1, choosed, i + 1);
		}
	}
	
	public static void main(String[] args) throws IOException {
		tokens = new StringTokenizer(input.readLine());
		N = Integer.parseInt(tokens.nextToken());
		M = Integer.parseInt(tokens.nextToken());
		for(int r=0; r<N; r++) {
			tokens = new StringTokenizer(input.readLine());
			for(int c=0; c<N; c++) {
				int map = Integer.parseInt(tokens.nextToken());
				if(map == 1) {
					Home.add(new int[] {r, c});
				}else if(map == 2) {
					Chicken.add(new int[] {r, c});
				}
			}
		}
		// 각 치킨집부터 집까지의 거리 구하기
		dis = new int[Chicken.size()][Home.size()];
		for(int c=0; c<Chicken.size(); c++) {
			for(int h=0; h<Home.size(); h++) {
				dis[c][h] = Math.abs(Chicken.get(c)[0] - Home.get(h)[0])
						  + Math.abs(Chicken.get(c)[1] - Home.get(h)[1]);
			}
		}		
		if(M > Chicken.size()) M = Chicken.size();
		combi(0, new int[M], 0);
		System.out.println(Min);
	}
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.js
// 도시 N*N
//   - 0 : 빈칸
//   - 1 : 집
//   - 2 : 치킨집
// 치킨 거리
//   - 집과 가장 가까운 치킨집 사이의 거리
//   - |r1 - r2| + |c1 - c2|
// 도시의 치킨 거리
//   - 모든 집의 치킨 거리의 합
// 가장 수익을 많이 낼 수 있는 치킨집 M개를 고르고 나머지는 모두 폐업
// => 출력
//    M개 고른 후 도시 치킨 거리의 최솟값
// => 풀이
//    - 모든 치킨집의 위치를 넣은 배열과, 모든 집 위치를 넣은 배열을 만듦
//    - 치킨집 중 m개 고르고 
//      -> 각 치킨집을 돌면서
//      -> 치킨집에서 각 집까지의 거리들을 측정
//      -> 각 집의 치킨 거리 갱신
//      => 도시 치킨 거리 갱신
//    - m===6 -> 치킨집갯수Cm * (m*50) = 1716 * 30 = 5만

let N, M;
let Map;
let Chicken = [];
let House = [];
let chickenLen;
let houseLen;
let MinCityDis = Number.MAX_SAFE_INTEGER;

const Loc = function(r, c) {
    this.r = r;
    this.c = c;
}

const measure = (selected) => {
    let dis = new Array(houseLen).fill(Number.MAX_SAFE_INTEGER);

    // 각 치킨집에서 모든 집을 돌면서 각 집의 치킨 거리 구하기
    selected.forEach(cIdx => {
        const cLoc = Chicken[cIdx];
        House.forEach((hLoc, hIdx) => {
            const d = Math.abs(cLoc.r - hLoc.r) + Math.abs(cLoc.c - hLoc.c);
            if(dis[hIdx] > d) {
                dis[hIdx] = d;
            }
        })
    })

    // 도시 치킨거리 갱신
    let cityDis = 0;
    dis.forEach(el => cityDis += el)
    MinCityDis = Math.min(MinCityDis, cityDis);
}

const combi = (start, nth, selected) => {
    if(nth === M) {
        // 각 치킨집 돌면서 모든 집의 치킨 거리 측정
        measure(selected);

        return;
    }

    for(let i=start; i<chickenLen; i++) {
        selected[nth] = i;
        combi(i+1, nth+1, [...selected]);
    }
}

const main = () => {
    const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

    [N, M] = input.shift().split(' ').map(Number);
    Map = input.map(row => row.split(' ').map(Number));

    // 모든 치킨집 위치와 집 위치 넣은 배열 만들기
    Map.forEach((row, r) => row.forEach((el, c) => {
        if(el === 1) House.push(new Loc(r, c));
        else if(el === 2) Chicken.push(new Loc(r, c));
    }))
    houseLen = House.length;
    chickenLen = Chicken.length;

    // 하고많은 치킨집중에 m개 골라보자
    combi(0, 0, new Array(M).fill(0));
    
    console.log(MinCityDis);
}

main();